### PR TITLE
[Enhancement]: block prefetch and locality [part 2]

### PR DESCRIPTION
**Pull request by badboynt1**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #9016

## What this PR does / why we need it:
group blocks to objects and prefetch object before read blocks

### CODE REVIEW

File: pkg/vm/engine/disttae/engine.go [hunk 1]
```diff
@@ -464,12 +464,12 @@ func (e *Engine) Hints() (h engine.Hints) {
 func (e *Engine) NewBlockReader(ctx context.Context, num int, ts timestamp.Timestamp,
 	expr *plan.Expr, ranges [][]byte, tblDef *plan.TableDef) ([]engine.Reader, error) {
 	rds := make([]engine.Reader, num)
-	blks := make([]catalog.BlockInfo, len(ranges))
+	blks := make([]*catalog.BlockInfo, len(ranges))
```
Comment (fengttt) [inline pkg/vm/engine/disttae/engine.go:467]: why do u need ptr?

**Review by fengttt — APPROVED** [review]

File: pkg/vm/engine/disttae/engine.go (hunk 1 above)
Comment (badboynt1) [reply pkg/vm/engine/disttae/engine.go:467]: to avoid copy of BlockInfo. better performance

File: pkg/vm/engine/disttae/tools.go [hunk 2]
```diff
@@ -1341,3 +1342,61 @@ func transferDecimal128val(a, b int64, oid types.T) (bool, any) {
...
+
+func groupBlocksToObjects(blocks []*catalog.BlockInfo, dop int) ([][]*catalog.BlockInfo, []int) {
+	var infos [][]*catalog.BlockInfo
+	objMap := make(map[string]int, 0)
+	lenObjs := 0
+	for i := range blocks {
+		block := blocks[i]
+		objName := block.MetaLocation().Name().String()
+		if idx, ok := objMap[objName]; ok {
+			infos[idx] = append(infos[idx], block)
```
Comment (nnsgmsone) [inline pkg/vm/engine/disttae/tools.go:1354]: Don't you need to sort here? Don't you need to ensure the order inside the object?

**Review by nnsgmsone — APPROVED** [review]

**Review by LeftHandCold — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Comment by sukki37** [comment]
@mergify unqueue

**Comment by guguducken** [comment]
@mergify refresh

**Comment by sukki37** [comment]
@mergify refresh

**Comment by guguducken** [comment]
@mergify requeue

**Pull request merged**